### ISSUE TITLE
Firewall: move Interface rule tabs to menu

### ISSUE THREAD

**Issue by fichtner** [opnsense/core#1928]
I think the firewall rules are the last legacy user of the tabbing. Several users noted that a lot of interfaces create an awkward tabbing experience and we already have a similar approach to interface editing in the menu already.

What do you think @AdSchellevis ? Feel free to reassign this.

**Comment by fichtner** [comment]
Some more info in #522,  display_top_tabs() is used for dhcp thingies as well, which are worth cleaning up too.

**Comment by mimugmail** [comment]
If you would do auto firewall rules for every local service you wouldnt need tabs et all. Sophos e.g. has only forwarding and it works great

**Comment by StrikerTwo** [comment]
I would also consider dropping the interface separation altogether and just having one big list of rules.

If I want to allow traffic to pass from one interface to the other and back (think LAN and DMZ) I have to switch between tabs (or menu items) to see all rules belonging to that flow.
Currently I tend to do as much as possible with floating rules because I can have them all in a single list.

I'd rather have some method of grouping rules that belong together than a separation by interface. An option to hide auto-created rules (NAT, RFC1918) would help to make it more clear as well.

**Comment by fichtner** [comment]
@mimugmail @StrikerTwo sounds like a long-term improvement, let's try to keep it simple in each iteration:

<img width="845" alt="screen shot 2017-11-19 at 10 55 31 am" src="https://user-images.githubusercontent.com/1915288/32989524-38ce3856-cd18-11e7-9796-b7d60fd5464d.png">

**Comment by AdSchellevis** [comment]
@fichtner keep in mind that with a lot of interfaces the plugging in the menu likely won't work properly. (the old page had a dropdown list for these circumstances)

**Comment by fichtner** [comment]
I guess that is why people probably rather want to see all of this interface-based stuff go away from navigation itself. either way many interfaces is a bit troublesome, at least from the menu we can expand and collapse and search via quick nav :)

**Comment by StrikerTwo** [comment]
@fichtner 
> sounds like a long-term improvement, let's try to keep it simple in each iteration

Yeah, sure. Should I open a new issue or do you keep a backlog somewhere?

**Comment by fichtner** [comment]
@StrikerTwo yes, I'm changing this ticket to kill all legacy tabs and put them in the menu instead

**Comment by AdSchellevis** [comment]
@fichtner I don't mind adding the navigation to the left menu, but I do think we should add the tabs too in that case, like we do in firmware for example. This isn't the best solution for everyone in my opinion, don't want to restructure the world either at the moment, but I'm really doubting we should remove the tabs completely now.

**Comment by fichtner** [comment]
This has been started a long time ago where I migrated all static tabs (e.g. VPN, System Settings) and wanted to kill tabbing completely. Since then, only interface-based magic remained marked as TODO for a long time and the menu system now more capable of emulating this. We have not added a single interface-based navigation page in MVC like this ever again so the issue is not a big one in UX, we lose the top tabs but still have the menu and the quick search now.

From the latest tab rework which I really like we now have two differing layouts which is highly confusing and counter intuitive. And I do consider this interface-based tabbing to be a mistake in design (we have not used this approach in our MVC code) and modelling to begin with so any step moving it out of the page content is a useful one. Menu-based design is also flexible in the sense that we can hide menus when they have no contents, e.g. router advertisement menu is not available if no static IPv6 config exists on an interface.

Now there is only one caller left: wireless status page. I will kill this one for sure. If we want to keep the firewall rules tabs, they will be the only legacy tabs and likely need that design brush-up for consistency. I'm fine with that too, also fine with moving ahead with this as late as 18.1, but we should do something of value in the old pages, not wait to replace them one fine day, which easily becomes a far fine day in the distant future.

**Comment by AdSchellevis** [comment]
Sorry, maybe my intentions weren't clear in my previous comment. 

I definitely don't think we should abandon the legacy pages until all the world around it has changed, I already spend way too much time in there for that. 

Removing the tabs for most cases is absolutely fine too.... but... the firewall page is really a corner-case, if you have more then 10-20 interfaces you will likely struggle a lot with the menu items on the left. At the moment we don't have a decent solution for that, which will really impact UX for people having larger installations. (the search menu works great, only not if you need to configure the interfaces one by one).

Maybe I can do a suggestion, I will add some tabs in the the code there is now which fits the styling of the new pages (and has similar support for multiple interfaces), then after 18.1 we look into better presentation options (can also be legacy for my concern, given the complexity underneath, but should work for larger installations too).
Is that something you can live with?

**Comment by fichtner** [comment]
Sure, should we make the menu entries searchable-only? Then I can revert the styling for firewall rules page.

This "issue" is already in place since we moved the interfaces [XXX] entries into the menu, firewall items will be slightly bigger. To this day, there were no raised question about this move so I thought it was a worthwhile task to finish.

But I do want to keep the modifications for DHCP and wireless.

**Comment by AdSchellevis** [comment]
My only concern are the firewall pages, people don't change the interfaces that much. DHCP and wireless I can live with, although I expect people to complain about that too at some point in time (because of the size of the left menu).

Maybe it's better to leave the tabs out of the firewall section for now and add something in which fits the styling, so you don't have to revert your commit and we can ditch display_top_tabs().
I'll take care of that for you.

**Comment by fichtner** [comment]
Ok thanks, I see your concern for firewall rules usage. The wider suggestion here was to show all rules there, or do some kind of selector like the categories for quicker usage. Floating rules break this paradigm a bit as their order is different in the final rules.

It's difficult to find an easy way out but maybe we can discuss this week after a bit of working with the new menu state. I don't mind circling back. Working on this is easy at first, yet working on the right improvements is certainly harder but worthwhile with a little help and review.

Thanks! :)

**Comment by StrikerTwo** [comment]
Can you give us some numbers? How many interfaces would the "new" menu support and how many installations are there that have more?
Can you do an ugly fix for the menu system if there is an excessive number of interfaces? A simple `if (cnt > x) WriteDropDown()` should be enough, no?

**Comment by fichtner** [comment]
I don't think drop down boxes in the menu are the answer, it would have to be some "..." entry that expands, but that is also custom glue. The easiest solution is to keep this away from the menu and provide a sane sort of filtering on the main page that is not tabbing.
